File: internal/explore/records.go
package explore

import (
	"fmt"

	"github.com/atyrode/babel/internal/disposition"
	"github.com/atyrode/babel/internal/frontier"
)

// The writes below share one shape, and it is the shape resumption needs.
//
// Each consults the resume ledger for the reference the worker emitted the
// item under. A hit means a prior attempt at this run already wrote the
// record, so the durable identifier is reused and nothing is written: that is
// what makes a re-run after an interruption neither lose nor duplicate
// committed state (§6.5). A miss writes the record and then binds the
// reference to it, in that order, because a binding pointing at a record that
// does not exist would be worse than a second attempt discovering the record
// is missing and writing it.
//
// A reference already bound to a record of another kind is an error rather
// than an overwrite. Records are immutable revisions (§4.7); a reference that
// changed its mind about what it names is a corrupt ledger.

// reuse resolves a reference through the ledger, requiring the recorded kind.
func reuse(committed map[string]Commit, ref string, want frontier.EntityType) (string, bool, error) {
	prior, ok := committed[ref]
	if !ok {
		return "", false, nil
	}
	if prior.Type != want {
		return "", false, fmt.Errorf("%w: reference %q names a %s, not a %s",
			ErrLedgerConflict, ref, prior.Type, want)
	}
	return prior.ID, true, nil
}

func (c *Controller) putHypothesis(st *state, stage Stage, runID string, committed map[string]Commit, cand Candidate) (string, bool, error) {
	if id, reused, err := reuse(committed, cand.Ref, frontier.EntityHypothesis); reused || err != nil {
		return id, reused, err
	}
	record, err := c.cfg.Frontier.CreateHypothesis(st.commit, frontier.HypothesisInput{
		RunID:   runID,
		Payload: cand.Hypothesis,
		// The dedup check runs before the write and travels with it, so the
		// warning and the candidate land in one transaction: a candidate
		// stored without the warning computed for it would be a suspicion
		// that existed only in a log. The candidate is written whatever the
		// heuristic said — never dropped, never reworded (#87).
		NearDuplicates: c.nearDuplicates(st, cand.Hypothesis.Statement),
	})
	if err != nil {
		return "", false, err
	}
	st.out.Duplicates = append(st.out.Duplicates, record.Duplicates...)
	return record.ID, false, c.bind(st, stage, cand.Ref, frontier.EntityHypothesis, record.ID)
}

func (c *Controller) putObservation(st *state, stage Stage, runID string, committed map[string]Commit, hypothesisID string, obs Observation) (string, bool, error) {
	id, reused, err := reuse(committed, obs.Ref, frontier.EntityObservation)
	if err != nil {
		return "", false, err
	}
	if reused {
		// Re-emitted on the resume path too. The claim's locators are the
		// worker's own, so they are the same on the attempt that wrote the
		// record and on the one that recognized it, and an interruption
		// between the write and its edges is repaired rather than permanent.
		c.mintEvidence(st, stage, runID, frontier.EntityObservation, id, obs.Claim.Evidence)
		return id, true, nil
	}
	record, err := c.cfg.Frontier.CreateObservation(st.commit, frontier.ObservationInput{
		HypothesisID:  hypothesisID,
		RunID:         runID,
		RecipeID:      obs.Recipe.ID,
		RecipeVersion: obs.Recipe.Version,
		Payload:       obs.Claim,
	})
	if err != nil {
		return "", false, err
	}
	c.mintEvidence(st, stage, runID, frontier.EntityObservation, record.ID, obs.Claim.Evidence)
	return record.ID, false, c.bind(st, stage, obs.Ref, frontier.EntityObservation, record.ID)
}

// putObjection records one challenger criticism, as an observation when it
// carries locators and as a contradicting candidate when it does not.
//
// The branch is the challenger's authority made mechanical: §4.3 forbids an
// evidence-free observation, and §5.4 permits a challenger to emit a new
// hypothesis, so an ungrounded criticism becomes an idea to investigate rather
// than a claim that has been established by being asserted.
func (c *Controller) putObjection(st *state, stage Stage, runID string, committed map[string]Commit, target string, obj Objection) (string, frontier.EntityType, bool, error) {
	if prior, ok := committed[obj.Ref]; ok {
		// A criticism recognized from a prior attempt gets its evidence
		// edges re-emitted for the same reason putObservation's does; the
		// branch below chose the kind then, and the ledger remembers it.
		if prior.Type == frontier.EntityObservation {
			c.mintEvidence(st, stage, runID, prior.Type, prior.ID, obj.Claim.Evidence)
		}
		return prior.ID, prior.Type, true, nil
	}
	if len(obj.Claim.Evidence) > 0 {
		record, err := c.cfg.Frontier.CreateObservation(st.commit, frontier.ObservationInput{
			HypothesisID:  target,
			RunID:         runID,
			RecipeID:      obj.Recipe.ID,
			RecipeVersion: obj.Recipe.Version,
			Payload:       obj.Claim,
		})
		if err != nil {
			return "", "", false, err
		}
		c.mintEvidence(st, stage, runID, frontier.EntityObservation, record.ID, obj.Claim.Evidence)
		return record.ID, frontier.EntityObservation, false,
			c.bind(st, stage, obj.Ref, frontier.EntityObservation, record.ID)
	}

	record, err := c.cfg.Frontier.CreateHypothesis(st.commit, frontier.HypothesisInput{
		RunID: runID,
		Payload: frontier.HypothesisPayload{
			Statement:         obj.Claim.Claim,
			OriginCues:        []string{"challenger objection grounded in " + string(obj.Grounds)},
			ProvisionalLabels: []string{"objection"},
			Notes:             obj.Claim.Category,
		},
	})
	if err != nil {
		return "", "", false, err
	}
	if _, err := c.cfg.Frontier.Link(st.commit, frontier.LinkInput{
		FromID: record.ID,
		ToID:   target,
		Type:   frontier.LinkContradicts,
		Note:   "challenger objection resting on " + string(obj.Grounds),
	}); err != nil {
		return "", "", false, err
	}
	return record.ID, frontier.EntityHypothesis, false,
		c.bind(st, stage, obj.Ref, frontier.EntityHypothesis, record.ID)
}

func (c *Controller) putFinding(st *state, stage Stage, runID string, committed map[string]Commit, con Consolidation, observationIDs []string) (frontier.Finding, bool, error) {
	if id, reused, err := reuse(committed, con.Ref, frontier.EntityFinding); err != nil {
		return frontier.Finding{}, false, err
	} else if reused {
		record, err := c.cfg.Frontier.Finding(st.commit, id)
		return record, true, err
	}
	record, err := c.cfg.Frontier.CreateFinding(st.commit, frontier.FindingInput{
		RunID:          runID,
		ObservationIDs: observationIDs,
		Payload:        con.Finding,
	})
	if err != nil {
		return frontier.Finding{}, false, err
	}
	return record, false, c.bind(st, stage, con.Ref, frontier.EntityFinding, record.ID)
}

func (c *Controller) putProposal(st *state, stage Stage, runID string, committed map[string]Commit, con Consolidation, findingID string) (string, bool, error) {
	ref := con.Ref + "/proposal"
	if id, reused, err := reuse(committed, ref, frontier.EntityProposal); reused || err != nil {
		return id, reused, err
	}
	record, err := c.cfg.Frontier.CreateProposal(st.commit, frontier.ProposalInput{
		RunID:      runID,
		FindingIDs: []string{findingID},
		Payload:    *con.Proposal,
	})
	if err != nil {
		return "", false, err
	}
	return record.ID, false, c.bind(st, stage, ref, frontier.EntityProposal, record.ID)
}

// bind records the reference-to-record binding a resumed attempt reads.
func (c *Controller) bind(st *state, stage Stage, ref string, kind frontier.EntityType, id string) error {
	return c.cfg.Ledger.Record(st.commit, st.opt.RunID, stage, Commit{
		Ref:  ref,
		Type: kind,
		ID:   id,
		At:   c.now(),
	})
}

// putDispositions records the next actions a job proposed against one record.
//
// Idempotence is the disposition store's rather than the resume ledger's. The
// ledger binds a worker reference to a frontier record and its Commit names a
// frontier entity kind; widening that kind to cover a second table would make
// every resume read ambiguous about what it is resolving. The store keys a
// run's proposal by (run, ref) instead, which is the same guarantee held one
// table over: a replayed result finds its own prior proposal and adds nothing.
//
// Each action is recorded on its own. One refused action — an unknown kind, an
// unverifiable repository — is a recorded failure for that action and leaves
// the others, because a job that proposed four next steps and got one wrong
// proposed three good ones.
func (c *Controller) putDispositions(st *state, stage Stage, runID string, record frontier.Ref, actions []ProposedAction) {
	for _, action := range actions {
		payload := disposition.Payload{Summary: action.Summary, Rationale: action.Rationale}
		if action.Kind == disposition.KindDraftIssue {
			anchor, err := disposition.VerifyAnchor(action.Workspace)
			if err != nil {
				st.fail(stage, FailureDisposition, c.now(), fmt.Errorf(
					"explore: draft-issue disposition %q: %w", action.Ref, err))
				continue
			}
			payload.Anchor = &anchor
		}
		proposed, err := c.cfg.Dispositions.Propose(st.commit, disposition.ProposeInput{
			Record:     record,
			Kind:       action.Kind,
			ProposedBy: frontier.Run(runID),
			Ref:        action.Ref,
			Payload:    payload,
		})
		if err != nil {
			st.fail(stage, FailureDisposition, c.now(), fmt.Errorf(
				"explore: persist disposition %q: %w", action.Ref, err))
			continue
		}
		st.out.Dispositions = append(st.out.Dispositions, proposed.ID)
	}
}

File: atyrode.babel/server/engine/records.ts
import {
  JOB_OUTPUT_FILES,
  MATERIAL_ROOT,
  MATERIAL_SESSIONS,
  type MaterialEntry,
} from "../../contract.ts";
import {
  REFUSALS,
  RESULT_SCHEMA,
  ResultRefusal,
  type Candidate,
  type Consolidation,
  type Evidence,
  type ExploreResult,
  type Objection,
  type QuestionDraft,
} from "../../machine/results.ts";
import { mintId, titleCell, type Row } from "./rows.ts";

/*
  AN ACCEPTED EXPLORATION, TURNED INTO THE ROWS IT CLAIMED.

  This is the other half of a settlement, and until it existed the first half was the whole of
  it: `settleSession` read the answer, checked every citation against the material and wrote a
  receipt, and the hypotheses, observations, findings and proposals the model had actually
  produced went nowhere. Every record this plugin held was imported Go-era history; nothing it
  ran ever added one. Ported from `v0.4.0:internal/explore/{records.go,stages.go,questions.go}`, whose
  `putHypothesis`/`putObservation`/`putFinding`/`putProposal`/`schedule`/`ask` are the reference
  for what one answer becomes.

  THE DEVELOPMENT PATH IS THE STRUCTURE, not a convention laid over it (§4.2). A candidate is a
  hypothesis; an observation hangs off exactly one hypothesis as its `parent_id` and carries the
  locators that make it evidence; a finding CONSOLIDATES observations and nothing else; a
  proposal ADDRESSES the finding or the hypothesis it answers. Those four sentences are the
  edges, in that direction, and they are the direction the crossing wrote every existing edge in
  (`tools/import.ts`: `consolidates` finding→observation, `addresses` proposal→finding and
  proposal→hypothesis, `cites` record→session).

  NOTHING HERE IS A RULING. A run may write records, edges, statuses and questions, and may not
  write a `dispositions` row: a ruling is the operator's alone, and a table a run could reach
  would make Babel an agent that agrees with itself.

  EVERY IDENTIFIER IS MINTED FROM THE RUN AND THE MODEL'S OWN HANDLE ({@link mintId}), which is
  what the Go tree's resume ledger did with a table: the second settlement of one run mints the
  same identifiers, every insert is `INSERT OR IGNORE` keyed by them, and a retry neither
  duplicates a record nor loses one. The handles are the answer's `ref` fields; a family prefix
  is part of the digest, so a candidate and an observation sharing a handle are still two rows.
*/

/** The payload shape a record's own JSON declares, as {@link RESULT_SCHEMA}'s version. */
const PAYLOAD_SCHEMA = Number(RESULT_SCHEMA.slice(RESULT_SCHEMA.lastIndexOf("/") + 1));

/** What a settlement needs to know about the run whose answer it is writing. */
export interface ExploreSettlement {
  readonly runId: string;
  /** ISO instant every row is stamped with: one settlement is one moment. */
  readonly at: string;
  /** The material this run was served, which is what a `cites` edge resolves a locator through. */
  readonly sessions: readonly MaterialEntry[];
}

/** The rows one answer becomes, keyed by the output file the ingest binds to each table. */
export type ExploreRows = Readonly<Record<string, readonly Row[]>>;

export interface ExploreWrite {
  readonly rows: ExploreRows;
  /**
   * What was recorded and dropped rather than refused. A disposal naming a handle the result
   * never declared loses only its scheduling note, so the run's records stand — the Go tree
   * warned for the same reason (`stages.go`: "the candidate the note was about does not exist").
   */
  readonly notes: readonly string[];
}

/**
 * Every row one accepted exploration claimed, or the refusal that stops the whole answer.
 *
 * A refusal here is the same kind of event as a schema refusal: the model answered, the
 * deployment paid, and the answer did not stand — so the caller writes the receipt at cost and
 * inserts NOTHING. It is all-or-nothing on purpose. A finding whose observations were dropped is
 * a consolidation of records that do not exist, which is exactly the shape §4.2 forbids, and
 * half a development path is worse than none: nobody can tell later which half is missing.
 */
export function exploreRows(
  result: ExploreResult,
  settlement: ExploreSettlement,
): ExploreWrite | { refusal: ResultRefusal } {
  const rows: Record<string, Row[]> = {
    [JOB_OUTPUT_FILES.records]: [],
    [JOB_OUTPUT_FILES.edges]: [],
    [JOB_OUTPUT_FILES.statusEvents]: [],
    [JOB_OUTPUT_FILES.questions]: [],
  };
  const notes: string[] = [];
  const writer = new Writer(rows, notes, settlement);
  try {
    for (const candidate of result.candidates) writer.candidate(candidate);
    for (const objection of result.objections) writer.objection(objection);
    for (const consolidation of result.consolidations) writer.consolidation(consolidation);
    writer.schedule(result);
    for (const question of result.questions) writer.question(question);
  } catch (error) {
    if (error instanceof ResultRefusal) return { refusal: error };
    throw error;
  }
  return { rows, notes };
}

/**
 * The selector of each session in the material, by every path a locator may name it with.
 *
 * It is the same three spellings `unservedLocator` admits, and for the same reason: the
 * prompt names the material's files relative to its root, a model may cite either spelling, and
 * a `cites` edge addresses the SESSION rather than the file — so the path has to resolve to the
 * selector the catalog holds.
 */
function selectors(sessions: readonly MaterialEntry[]): Map<string, string> {
  const out = new Map<string, string>();
  for (const entry of sessions) {
    out.set(entry.file, entry.selector);
    out.set(`${MATERIAL_SESSIONS}/${entry.file}`, entry.selector);
    out.set(`${MATERIAL_ROOT}/${MATERIAL_SESSIONS}/${entry.file}`, entry.selector);
  }
  return out;
}

/**
 * A durable identifier of one record family, as `RecordIdSchema` shapes them.
 *
 * A handle is resolved BY FAMILY rather than by looking like an identifier: a consolidation
 * citing `hyp_…` from the brief has skipped the development path just as surely as one citing a
 * candidate this result declared, and a check that admitted any record id would let it through.
 */
const recordId = (family: string): RegExp => new RegExp(`^${family}_[0-9a-f]{8,64}$`, "u");
const HYPOTHESIS_ID = recordId("hyp");
const OBSERVATION_ID = recordId("obs");

class Writer {
  private readonly hypotheses = new Map<string, string>();
  private readonly observations = new Map<string, string>();
  private readonly sessions: Map<string, string>;

  constructor(
    private readonly rows: Record<string, Row[]>,
    private readonly notes: string[],
    private readonly settlement: ExploreSettlement,
  ) {
    this.sessions = selectors(settlement.sessions);
  }

  /** One candidate: the hypothesis, its observations, and the change it proposes for itself. */
  candidate(candidate: Candidate): void {
    const id = this.mint("hyp", candidate.ref);
    this.hypotheses.set(candidate.ref, id);
    this.record(id, "hypothesis", candidate.hypothesis.statement, candidate.hypothesis);
    // THE LIFECYCLE STARTS AT UNTRIAGED, written with the record rather than inferred from its
    // absence: `status_events` is the history the frontier reads a standing off, and a
    // hypothesis with no event at all is one no listing can rank or defer.
    this.status(id, 0, "untriaged", "");
    for (const observation of candidate.observations) {
      const child = this.mint("obs", observation.ref);
      this.observations.set(observation.ref, child);
      this.record(child, "observation", observation.claim.claim, observation.claim, {
        parentId: id,
        recipe: observation.recipe,
      });
      this.cites(child, "observation", observation.claim.evidence);
    }
    const remedy = candidate.remedy;
    if (remedy === undefined) return;
    // §4.5's candidate proposal: the change this candidate asks for, addressing the claim beside
    // it and no finding. It runs after the hypothesis because it names it.
    const proposal = this.mint("pro", remedy.ref);
    this.record(proposal, "proposal", remedy.proposal.title, remedy.proposal);
    this.edge("addresses", proposal, "proposal", id, "hypothesis", 0, "");
  }

  /**
   * One challenger criticism: an observation when it carries locators, and a contradicting
   * candidate when it does not.
   *
   * The branch is §5.4's authority made mechanical, and it is the Go tree's (`putObjection`):
   * §4.3 forbids an evidence-free observation, so an ungrounded criticism becomes an idea to
   * investigate rather than a claim established by being asserted.
   */
  objection(objection: Objection): void {
    const target = this.hypothesis(objection.hypothesis, "objection");
    if (objection.claim.evidence.length > 0) {
      const id = this.mint("obs", objection.ref);
      this.observations.set(objection.ref, id);
      this.record(id, "observation", objection.claim.claim, objection.claim, {
        parentId: target,
        recipe: objection.recipe,
      });
      this.cites(id, "observation", objection.claim.evidence);
      return;
    }
    const id = this.mint("hyp", objection.ref);
    this.hypotheses.set(objection.ref, id);
    this.record(id, "hypothesis", objection.claim.claim, {
      statement: objection.claim.claim,
      origin_cues: [`challenger objection grounded in ${objection.grounds}`],
      provisional_labels: ["objection"],
      novelty: 0,
      priority: 0,
      notes: objection.claim.category,
    });
    this.status(id, 0, "untriaged", "");
    this.edge(
      "contradicts",
      id,
      "hypothesis",
      target,
      "hypothesis",
      0,
      `challenger objection resting on ${objection.grounds}`,
    );
  }

  /** One consolidation: what recurs across observations, and the change it asks for. */
  consolidation(consolidation: Consolidation): void {
    const supports = consolidation.observations.map((ref) => this.observation(ref));
    const id = this.mint("fnd", consolidation.ref);
    this.record(id, "finding", consolidation.finding.title, consolidation.finding);
    for (const [position, support] of supports.entries()) {
      this.edge("consolidates", id, "finding", support, "observation", position, "");
    }
    const proposal = consolidation.proposal;
    if (proposal === undefined) return;
    const proposalId = this.mint("pro", `${consolidation.ref}/proposal`);
    this.record(proposalId, "proposal", proposal.title, proposal);
    this.edge("addresses", proposalId, "proposal", id, "finding", 0, "");
  }

  /**
   * What this run said it would not develop. §5.2 keeps it to scheduling: a deferred or rejected
   * candidate keeps its record and its wording, and only the lifecycle history changes.
   *
   * A rejection outranks a deferral of the same candidate — the stronger verdict is the one the
   * stage reached — and each candidate takes exactly one event, at `seq` 1, after the
   * `untriaged` its creation wrote.
   */
  schedule(result: ExploreResult): void {
    const scheduled = new Map<string, { status: string; reason: string }>();
    for (const disposal of result.deferred) {
      const id = this.scheduled(disposal.hypothesis, "deferral");
      if (id === "") continue;
      scheduled.set(id, { status: "deferred", reason: disposal.reason });
    }
    for (const disposal of result.rejected) {
      const id = this.scheduled(disposal.hypothesis, "rejection");
      if (id === "") continue;
      scheduled.set(id, { status: "rejected", reason: disposal.reason });
    }
    for (const [id, verdict] of scheduled) this.status(id, 1, verdict.status, verdict.reason);
  }

  /**
   * One thing the corpus could not settle and a person can (§4.8).
   *
   * A run may raise one at all because it authorizes nothing: it is a request that somebody else
   * authorize something, so accepting it costs the ledger no authority. Two things follow, and
   * both are the Go tree's (`questions.go`). The KIND is always `acquire-context` — the ledger's
   * other kinds are conditions only the ledger's own machinery can see. The CLASS is derived
   * from the work rather than declared: a question blocking a candidate this run wrote ranks
   * above one that satisfies curiosity, and a self-graded class would make every question
   * blocking within a week.
   *
   * The subjects stay strings in the payload. A run does not mint identity (§4.8 puts entity
   * creation behind an attributed operator act), so a subject naming something the ledger does
   * not hold is a gap in the ledger to read, not an `entities` row this run may create.
   */
  question(draft: QuestionDraft): void {
    const blocked =
      draft.hypothesis === "" ? "" : (this.hypotheses.get(draft.hypothesis) ?? draft.hypothesis);
    this.rows[JOB_OUTPUT_FILES.questions]?.push({
      id: mintId("qst", this.settlement.runId, draft.ref),
      kind: "acquire-context",
      class: blocked === "" ? "curiosity" : "blocking",
      text: draft.prompt,
      why: draft.why_asked,
      dedupe_key: null,
      raised_by_kind: "run",
      raised_by_id: this.settlement.runId,
      payload: JSON.stringify({
        schema: PAYLOAD_SCHEMA,
        prompt: draft.prompt,
        why_asked: draft.why_asked,
        subjects: draft.subjects,
        predicates: draft.predicates,
        sensitivity: "routine",
        expected_authority: "operator",
        work: blocked === "" ? [] : [{ kind: "hypothesis", id: blocked, blocking: true }],
      }),
      created_at: this.settlement.at,
    });
  }

  // -------------------------------------------------------------------------- resolution

  private mint(prefix: string, ref: string): string {
    return mintId(prefix, this.settlement.runId, ref);
  }

  /** A handle naming a candidate: one this result declared, or a durable id the brief named. */
  private hypothesis(ref: string, what: string): string {
    const local = this.hypotheses.get(ref);
    if (local !== undefined) return local;
    if (HYPOTHESIS_ID.test(ref)) return ref;
    throw new ResultRefusal(
      REFUSALS.unknownReference,
      `the ${what} names ${ref}, which is neither a candidate this result declared nor a hypothesis identifier`,
    );
  }

  /**
   * A handle naming an observation, with §4.2's path enforced.
   *
   * A consolidation resting on a CANDIDATE is the mandatory path skipped rather than a missing
   * reference, and the two are separate refusals because they are separate mistakes: one says
   * the synthesizer cited a guess as if it were evidence, the other says it cited nothing at all.
   */
  private observation(ref: string): string {
    const local = this.observations.get(ref);
    if (local !== undefined) return local;
    if (this.hypotheses.has(ref) || HYPOTHESIS_ID.test(ref)) {
      throw new ResultRefusal(
        REFUSALS.developmentPath,
        `${ref} is a candidate hypothesis, not a locator-backed observation`,
      );
    }
    if (OBSERVATION_ID.test(ref)) return ref;
    throw new ResultRefusal(REFUSALS.unknownReference, `no observation ${ref}`);
  }

  /** A handle a disposal named, or the empty string with the note saying it was dropped. */
  private scheduled(ref: string, what: string): string {
    const local = this.hypotheses.get(ref);
    if (local !== undefined) return local;
    this.notes.push(
      `the ${what} of ${ref} names no candidate this run declared, so it was dropped`,
    );
    return "";
  }

  // -------------------------------------------------------------------------- rows

  private record(
    id: string,
    kind: string,
    title: string,
    payload: unknown,
    of?: { parentId?: string; recipe?: { id: string; version: number } },
  ): void {
    this.rows[JOB_OUTPUT_FILES.records]?.push({
      id,
      kind,
      // A NEW RECORD IS ITS OWN ROOT AT SEQUENCE ZERO. A correction is a supersession — the
      // table refuses an update by trigger — so nothing a run writes ever revises a row.
      root_id: id,
      supersedes_id: null,
      seq: 0,
      parent_id: of?.parentId ?? null,
      run_id: this.settlement.runId,
      // ONLY AN OBSERVATION CARRIES A RECIPE, which is what the answer attributes and what the
      // crossing wrote (`tools/import.ts`: the recipe columns are an observation's alone). A
      // hypothesis, a finding and a proposal are reached THROUGH a method and are not products
      // of one, and naming the run's first recipe on them would be an attribution nobody made.
      recipe_id: of?.recipe?.id ?? null,
      recipe_version: of?.recipe?.version ?? null,
      actor_kind: "run",
      actor_id: this.settlement.runId,
      title: titleCell(title),
      created_at: this.settlement.at,
      payload: JSON.stringify({ schema: PAYLOAD_SCHEMA, ...(payload as Record<string, unknown>) }),
    });
  }

  private status(recordId: string, seq: number, status: string, reason: string): void {
    this.rows[JOB_OUTPUT_FILES.statusEvents]?.push({
      id: mintId("ste", this.settlement.runId, `${recordId}|${String(seq)}`),
      record_id: recordId,
      seq,
      status,
      run_id: this.settlement.runId,
      actor_kind: "run",
      actor_id: this.settlement.runId,
      reason: reason === "" ? null : reason,
      recorded_at: this.settlement.at,
    });
  }

  /**
   * The sessions one claim reached for, as edges.
   *
   * The edge carries the SESSION and nothing about where inside it: a citation's line and its
   * note live in the record's own payload, which is what makes a claim readable on a hub that
   * cannot open the conversation it names (`store/store.ts`, `citations`). Only supporting
   * evidence mints an edge — counter-evidence is the claim's own honesty about itself, and
   * counting it as material reached for would inflate every corroboration measure with it.
   *
   * A locator whose path the material does not name cannot be reached here: the caller has
   * already refused the whole answer for it, which is the only reason this can skip silently.
   */
  private cites(fromId: string, fromKind: string, evidence: readonly Evidence[]): void {
    for (const [position, cited] of evidence.entries()) {
      const selector = this.sessions.get(cited.locator.path);
      if (selector === undefined) continue;
      this.edge("cites", fromId, fromKind, selector, "session", position, cited.note);
    }
  }

  private edge(
    kind: string,
    fromId: string,
    fromKind: string,
    toId: string,
    toKind: string,
    position: number,
    note: string,
  ): void {
    this.rows[JOB_OUTPUT_FILES.edges]?.push({
      id: mintId("edg", this.settlement.runId, `${kind}|${fromId}|${toId}`),
      kind,
      from_kind: fromKind,
      from_id: fromId,
      to_kind: toKind,
      to_id: toId,
      position,
      note: note === "" ? null : note,
      actor_kind: "run",
      actor_id: this.settlement.runId,
      created_at: this.settlement.at,
    });
  }
}
